### PR TITLE
Version 3

### LINKED ISSUE #30: use camelCase instead of snake_case
for v3.0.0

### LINKED ISSUE #35: Use 100% js implementation of Blob
if possible.

### LINKED ISSUE #38: Error when importing nhost-js-sdk in react-native
I'm trying to configure auth in an Expo (v.39.0.2) project but am running into issues when importing the nhost-js-sdk package. See picture below. The `esm` module do exists in the node_modules folder and I've tried the instructions several times. 
Any ideas? Looks like a dependency on the `blob` package.

nhost_error

### LINKED ISSUE #48: Better error message from v2 to v3
ex if `base_url` is used, tell that the new variable name is `baseURL` etc.

### PR DESCRIPTION
**Pull request by elitan**
Fixes #30
Fixes #35
Fixes #38
Fixes #48
Fixes #31 (reimplementation)

### CODE REVIEW

**Review by nunopato — APPROVED** [review]

**Comment by elitan** [comment]
> @elitan expo-secure-storage only allows alphanumeric characters and "-", "_", "." - i.e. changing the key used to store the refreshToken in expo-secure-storage to nhost:refreshToken throws an error. I suggest to change it to camelcase, i.e. nhostRefreshToken. I did it in my local nhost cli dev environment and it made everything work again, plus I doubt it will create any other problems for any other platform.

Done.

File: src/Auth.ts [hunk 1]
```diff
@@ -339,206 +194,430 @@ export default class Auth {
...
+    if (this.useCookies) return null;
+
+    return {
+      Authorization: `Bearer ${this.currentSession.getSession()?.jwt_token}`,
+    };
+  }
+
+  private _autoLogin(refreshToken: string | null): void {
+    if (this.ssr) {
+      this._clearSession();
```
Comment (elitan) [inline src/Auth.ts:514]: Skip this one on SSR. Will only show an error that it' can't clear the session.

File: src/Auth.ts [hunk 2]
```diff
@@ -339,206 +194,430 @@ export default class Auth {
...
 
-    // set refresh token
-    if (!this.use_cookies) {
-      await this.setItem("refresh_token", res.data.refresh_token);
+    this.currentSession.clearSession();
+    this._removeItem("nhostRefreshToken");
+  }
+
+  private async _setSession(session: types.Session) {
+    this.currentSession.setSession(session);
```
Comment (elitan) [inline src/Auth.ts:590]: early exit if no sessions exist, such as after register() when auto-activate user is `false`.

File: src/Auth.ts (hunk 1 above)
Comment (nunopato) [reply src/Auth.ts:514]: Just to be sure, you want to remove `this._clearSession();` but keep the `return` when `this.ssr`?

File: src/Auth.ts (hunk 2 above)
Comment (nunopato) [reply src/Auth.ts:590]: @elitan what do you mean by `no sessions exist`? `session` here cannot be `null`.

Comment (elitan) [reply src/Auth.ts:590]: if the setting "auto activate account on regsitration" is `false` and you register a new user the `session` will be `null` during registration.

File: src/Auth.ts (hunk 1 above)
Comment (elitan) [reply src/Auth.ts:514]: Yes

**Comment by elitan** [comment]
`authStateChanged` is only called on ` _setSession`. 

`authStateChanged` must also be called on `logout()`.

if logout is called twice in a row `authStateChanged` should only be called once since the auth state changes only once.

File: src/Auth.ts (hunk 2 above)
Comment (nunopato) [reply src/Auth.ts:590]: @elitan you're right, I missed those cases

File: src/Auth.ts (hunk 1 above)
Comment (nunopato) [reply src/Auth.ts:514]: 👍

**Pull request merged**